src/Pos/Delegation/Cede/Holders.hs

```diff
@@ -0,0 +1,67 @@
+-- | This module provides implementations of "MonadCede" based on
+-- either pure DB or DB+hashmap access.
+
+module Pos.Delegation.Cede.Holders
+       ( DBCede
+       , runDBCede
+
+       , MapCede
+       , runMapCede
+       , evalMapCede
+       ) where
+
+import           Universum
+
+import           Control.Lens                 (at, (%=))
+import           Control.Monad.Trans.Identity (IdentityT (..))
+import           Data.Coerce                  (coerce)
+import qualified Data.HashMap.Strict          as HM
+import qualified Ether
+
+import           Pos.DB.Class                 (MonadDBRead)
+import           Pos.Delegation.Cede.Class    (MonadCede (..), MonadCedeRead (..))
+import           Pos.Delegation.Cede.Types    (CedeModifier, DlgEdgeAction (..),
+                                               dlgEdgeActionIssuer)
+import qualified Pos.Delegation.DB            as DB
+import           Pos.Types                    (addressHash)
+import           Pos.Util.Util                (ether)
+
+----------------------------------------------------------------------------
+-- Pure database-only holder
+----------------------------------------------------------------------------
+
+data DBCedeTag
+
+type DBCede = Ether.TaggedTrans DBCedeTag IdentityT
+
+runDBCede :: DBCede m a -> m a
+runDBCede = coerce
+
+instance MonadDBRead m => MonadCedeRead (DBCede m) where
+    getPsk = DB.getPskByIssuer . Right
+
+----------------------------------------------------------------------------
+-- DB + CedeModifier resolving
+----------------------------------------------------------------------------
+
+-- | Monad transformer that holds extra layer of modifications to the
+-- underlying set of PSKs (which can be empty if you want).
+type MapCede = Ether.LazyStateT' CedeModifier
+
+runMapCede :: CedeModifier -> MapCede m a -> m (a, CedeModifier)
+runMapCede = flip Ether.runLazyStateT
+
+evalMapCede :: Monad m => CedeModifier -> MapCede m a -> m a
+evalMapCede = flip Ether.evalLazyStateT
+
+instance MonadDBRead m => MonadCedeRead (MapCede m) where
+    getPsk iPk =
+        ether $ use (at iPk) >>= \case
+            Nothing                -> lift $ DB.getPskByIssuer $ Right iPk
+            Just (DlgEdgeDel _)    -> pure Nothing
+            Just (DlgEdgeAdd psk ) -> pure (Just psk)
+
+instance MonadDBRead m => MonadCede (MapCede m) where
```

Comment:
`MonadCedeRead m` is more explicit reflect the hierarchy than `MonadDBRead m`.
and it give more flexibility (you can stack several `MapCede`)
(same for
`instance MonadDBRead m => MonadCedeRead (MapCede m) where`)
but it's cosmetic